### ISSUE TITLE
Expression evaluation ..

### ISSUE THREAD

**Issue by skx** [skx/evalfilter#21]
The implementation of expressions, as handled by the `if` statement, is not as clean or reliable as it could be. Specifically it started with the expectation it would only handle statements of the form:

* `if ( Field operator Field|Value ) { ..`

Later it was updated to allow single-argument tests:

* `If ( Field ) { ... `
* `if ( FunctionCall( arg1, ... argN ) ) { .. `

Now I've grafted on support for `&&`, along with `||`, but it is __not__ a general purpose expression parser, and will not allow you to write full expressions such as these:

* `if ( Day == "Monday" && ( Hour <= 9 || Hour >= 17 ) ) { ..`
* `if ( 1 + Count >= 10 ) { ...`
* `if ( ! Field ) { ... `

It should.  Actually parsing expressions, via the golang lexer/ast would be a perfect way out of the problem:

**Comment by skx** [comment]
See second example here:

* https://rosettacode.org/wiki/Arithmetic_Evaluator/Go

That uses the golang parser.  If we go down that road, add support for `Foo()` and `Field` we're good.

It means our `IfOperation` will become:

```
// IfOperation holds state for the `if` operation
type IfOperation struct {

	// Expression contains the expression to evaluate.
	Expression string

	// Operations to be carried out if the statement matches.
	True []Operation

	// Operations to be carried out if the statement does not.
	False []Operation
}
```

i.e. We lose the `Left` + `Right` + `Op`.

Parsing that on our side will be simple, then we sub-parse via go's built-in parser.

**Comment by skx** [comment]
Converting this was pretty simple, but revealed a whole bunch of problems:

* I changed the `if` structure to store the expression like so
  * `Expression []Argument`
* Then in the evaluation function I expanded these objects.
  * ` '[%!s(*runtime.FieldArgument=&{Username}) %!s(*runtime.OperatorArgument=&{==}) %!s(*runtime.StringArgument=&{bin}) %!s(*runtime.OperatorArgument=&{||}) %!s(*runtime.FieldArgument=&{Username}) %!s(*runtime.OperatorArgument=&{==}) %!s(*runtime.StringArgument=&{sys})]'`
    * Became `"mosquitto" == "bin" || "mosquitto" == "sys" `
* That worked but only because I started quoting the values returned from field-lookups.

In short there's a mismatch between how I treat objects, and how the golang parser wants to treat input.  I couldn't add `if ( Channel =~ "_alerts" )` either.  Because that's not a native operation.  Instead I'd have to write a function `contains( $src, $substring)` or `match( $src, $regexp)`

Not impossible, but given the limited gain, the fact nobody has used the library (?) and complained about the expression parser I think this is now "won't fix".

To solve the problem without this overhead I'd need to write a proper parser.  Not impossible, but overkill unless/until it is completely necessary.
